kernel: rtl8367, rtl8367b: use realtek,extif property

namiltd opened the pull request:
Use realtek,extif property instead of realtek,extif0 to extif2 by extending it with the cpu_port parameter.
Thanks to this, similarly to DSA drivers, we only operate on cpu_port, the extif number is automatically calculated (according to tables from https://github.com/openwrt/openwrt/pull/15033).
There are routers that use different versions of chips interchangeably, e.g. RTL8367S or RTL8367S-VB. The external ports have different exif assignments for these different chips. By being dependent on cpu_port, they can have the same settings even though they programmatically use different extif number.

Refers to:
- ar7242_tplink_tl-wr2543-v1
- mt7620a_iodata_wn-ac733gr3
- mt7620a_tplink_archer-c2-v1
- mt7620a_tplink_archer-c5-v4
- mt7620a_tplink_ec220-g5-v2
- mt7620a_zyxel_keenetic-viva
- netgear-wndap620
- netgear-wndap660
- rt3662_asus_rt-n56u
- rt3662_edimax_br-6475nd
- rt3662_samsung_cy-swr1100
- rt3662_dlink_dir-645
- rt3883_belkin_f9k1109v1

Tested on: 
- mt7620a_tplink_archer-c5-v4
```
[    3.712586] rtl8367b rtl8367s: using MDIO bus 'mdio'
[    3.718694] rtl8367b rtl8367s: RTL8367S chip found (num:6367 ver:00a0)
[    4.089434] rtl8367b rtl8367s: cpu_port:7, assigned to extif2
```

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8366_smi.h:
```diff
@@ -64,6 +64,7 @@ struct rtl8366_smi {
 	u8			dbg_vlan_4k_page;
 #endif
 	u32			phy_id;
+	u8			chip;
```
Why don't you use a enum for this? I think that would make the code much more readable.

namiltd replied:
I was thinking about it, but it would require the definition of this enum in rtl8366_smi.h, if you think it would be better, I can do it.

hauke replied:
For me it gets confusing because you are not using normal numbers, but also shift this variable. 

I think it is no problem to define a enum in this file.

namiltd replied:
Ok, I will do it in the evening, I will also change the chip to rtl8367b_chip to make it clear what it concerns.

namiltd replied:
Done

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8367.c:
```diff
@@ -1077,21 +1077,37 @@ static int rtl8367_led_blinkrate_set(struct rtl8366_smi *smi, unsigned int rate)
-		return -EINVAL;
+	cpu_port = be32_to_cpup(prop++);
+	switch (cpu_port) {
+	case RTL8367_CPU_PORT_NUM-1:
```
Please add a space:
```suggestion
	case RTL8367_CPU_PORT_NUM - 1:
```

namiltd replied:
Okay, I'll fix it.

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8367.c:
```diff
@@ -1077,21 +1077,37 @@ static int rtl8367_led_blinkrate_set(struct rtl8366_smi *smi, unsigned int rate)
 
 	prop = of_get_property(smi->parent->of_node, name, &size);
-	if (!prop)
-		return rtl8367_extif_init(smi, id, NULL);
```
You removed the support without configuration. Are we using configurations in all device tree files using the realtek driver?

namiltd replied:
The (!prop) case is handled in line 1091

namiltd replied:
The (!prop) case is handled in line 1091
There is a configuration for all devices with rtl8367

namiltd committed:
kernel: rtl8367: use realtek,extif property

Use realtek,extif property instead of realtek,extif0 and realtek,extif1
by extending it with the cpu_port parameter.
The extif number is automatically calculated based on cpu_port.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/15749
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8366_smi.h:
```diff
@@ -23,6 +23,25 @@ struct dentry;
+typedef enum rtl8367b_chip_e {
+	RTL8367B_CHIP_UNKNOWN = RTL8367B_FAMILY_UNKNOWN << RTL8367B_FAMILY_SHIFT,
+	RTL8367B_CHIP_UNKNOWN_RLVID0 = RTL8367B_FAMILY_B_RLVID0 << RTL8367B_FAMILY_SHIFT,
+	RTL8367B_CHIP_UNKNOWN_RLVID1 = RTL8367B_FAMILY_B_RLVID1 << RTL8367B_FAMILY_SHIFT, 
```
I do not like that the enum uses these strange values. 
I would prefer when the enum would have some normal numbers and is not used with shifts. 

When I see this code:
```
switch (smi->rtl8367b_chip >> RTL8367B_FAMILY_SHIFT) {
```
Maybe just store the family here.

namiltd replied:
I simplified the code, I no longer use RTL8367B_FAMILY_SHIFT.

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -1560,32 +1552,50 @@ static int rtl8367b_detect(struct rtl8366_smi *smi)
 
 	if (!chip_name) {
+		rlvid = (chip_ver >> RTL8367B_CHIP_VER_RLVID_SHIFT) &
+			RTL8367B_CHIP_VER_RLVID_MASK;
```
The driver will always stop probing when it reaches this code. 
The `rtl8367b_init_regs` function makes use of some enum values only st here.

The old code in `rtl8367b_init_regs` would also trigger for `RTL8367B_CHIP_RTL8367RB` and `RTL8367B_CHIP_RTL8367R_VB`. Are you sure this is not used to distinguish them more detailed?

namiltd replied:
This code only runs when !chip_name. RTL8367B_CHIP_RTL8367RB and RTL8367B_CHIP_RTL8367R_VB is done in the earlier code (a few lines above).

namiltd replied:
This code only runs when !chip_name. Detection of RTL8367B_CHIP_RTL8367RB and RTL8367B_CHIP_RTL8367R_VB is done in the earlier code (a few lines above). Code was tested for RTL8367B_CHIP_RTL8367S,
Assignments are the some as in https://github.com/openwrt/openwrt/pull/15033#issue-2217119665

hauke replied:
In the new code it was only executed for unsupported chips, in the old code it was also executed for some supported chips, see:
https://github.com/openwrt/openwrt/pull/15749/files?diff=split&w=0#r1693980451

hauke replied:
I looked a bit closer at the code and it should be fine if you remove the code for `rtl8367r_vb_initvals_0` and the code setting `RTL8367B_CHIP_UNKNOWN_RLVID0` and `RTL8367B_CHIP_UNKNOWN_RLVID1` as they did not work before too.

I overlooked before that the `rlvid` was calculated based on the `chip_ver`.

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -730,35 +731,23 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
+		initvals = rtl8367r_vb_initvals_1;
+		count = ARRAY_SIZE(rtl8367r_vb_initvals_1);
+		break;
+	case RTL8367B_CHIP_RTL8367S ... RTL8367B_CHIP_RTL8367RB_VB:
```
I do not like the `...` in the case statement, otherwise this change looks fine.
Please mention all options so it does not break in case someone adds a new enum option in the middle.

namiltd replied:
"..." removed

namiltd replied:
"..." syntax removed
The default case is basically run for RTL8367B_CHIP_UNKNOWN

namiltd committed:
kernel: rtl8367b: remove rlvid analysis

Remove the rlvid analysis because for the rtl8367b family chips supported
by the driver (rtl8367rb and rtl8367r-vb), rlvid is always equal to 1.
So the code for rlvid equal to 0 is completely unnecessary.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/15749
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd committed:
kernel: rtl8367b: use realtek,extif property

Use realtek,extif property instead of realtek,extif0 to extif2
by extending it with the cpu_port parameter.
The extif number is automatically calculated based on cpu_port.

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/15749
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
In order to reduce and optimize the code, I completely removed rlvid analysis.
Routers supported by openwrt with chips from the rtl8367b family are only in versions with rlvid equal to 1. (RTL8367R-VB, RTL8367RB and RTL8363SB)

hauke commented on target/linux/generic/files/drivers/net/phy/rtl8367b.c:
```diff
@@ -730,39 +516,24 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
+	case RTL8367B_CHIP_RTL8363SB:
+	case RTL8367B_CHIP_RTL8367RB:
+	case RTL8367B_CHIP_RTL8367R_VB:
+		initvals = rtl8367b_initvals;
```
I think you have to add the check for the `rlvid` here. 
The old code uses `rtl8367c_initvals` for `chip_num == 0x6367` which the new code does in the same way. 
If it is not  `chip_num == 0x6367`  it was checking the `rlvid`. 
If you are not sure that we do not support such chips I would prefer that you add the check here. 

I do not know Realtek product development in detail, but it is common to have multiple chip revisions, the later ones are fixing bugs and need less workarounds in the driver. One chip is often sold under different names for different markets. I assume that all the 3 product names listed here are using the same chip with different fusing and/or packaging and there are 2 revisions which need different initialization.

namiltd replied:
rlvid was tested ror chip_num != 0x6367
It could have three possibilities:
- be equal to 1
- be equal to 0
- other

For 1 the supported chips are: RTL8363SB, RTL8367RB and RTL8367R-VB (chip_ver = 0x**1**000 and 0x**1**010)
For <>0 and 1: an error was generated in rtl8367b_detect (unknow rlvid)
For 0: There are no routers that could belong here.
<body>

 Family RTL8367B | chip | chip_ver | rlvid
-- | -- | -- | --
rt3883_belkin_f9k1109v1 | RTL8367R-VB | 0x1010 | 1
mt7620a_tplink_archer-c2-v1 | RTL8367RB | 0x1000 | 1 
mt7620a_zyxel_keenetic-viva | RTL8367RB | 0x1000 | 1 
mt7620a_iodata_wn-ac733gr3 | RTL8367RB | 0x1000  | 1
rt3662_dlink_dir-645 | RTL8367RB | 0x1000 | 1
Netgear-wndap620 | RTL8363SB | 0x1000 | 1
Netgear-wndap660 | RTL8363SB | 0x1000 | 1
</body>

namiltd replied:
Because I am not sure about RTL8363SB.
I found various information on the Internet, one that has chip_ver=0x6000 and chip_num=0x1000 (https://www.spinics.net/lists/netdev/msg909306.html) and in https://github.com/cgoder/openwrt_rtk/blob/ d237e45ee480ee7e75c80a4c61e1a335a0a77755/rtk_openwrt_sdk/target/linux/rtkmipsel/files/drivers/net/rtl819x/rtl8367r/rtk_api.c#L77C1-L78C5 yet another code.
To be sure not to miss any version of the circuits, I go back to the previous code

namiltd replied:
To be sure I would need to have system logs for Netgear-wndap620 and Netgear-wndap660 for this code

namiltd replied:
I found the log https://forum.openwrt.org/t/wndap660-kernel-panic-during-installation/57770 According to it, the switch is detected as RTL8367RB, so rlvid is equal to 1. So the code without rlvid is correct.

@hauke What do you think, is it better to use a simpler code without rlvid or the code with RTL8367B_CHIP_UNKNOWN_RLVID0 and RTL8367B_CHIP_UNKNOWN_RLVID1 but they are not currently used in any of the routers. Do "spare" types make sense?

chunkeey commented:
Hmm, sure.

```

[   14.537819] rtl8367b plb:opb:rtl8367b: using MDIO bus 'emac_mdio'
[   14.611507] rtl8367b plb:opb:rtl8367b: RTL8363SB chip found (num:6000 ver:1000, mode:1000)
[   15.790835] rtl8367b plb:opb:rtl8367b: cpu_port:5, assigned to extif0
[...]

  _______                     ________        __
 |       |.-----.-----.-----.|  |  |  |.----.|  |_
 |   -   ||  _  |  -__|     ||  |  |  ||   _||   _|
 |_______||   __|_____|__|__||________||__|  |____|
          |__| W I R E L E S S   F R E E D O M
 -----------------------------------------------------
 OpenWrt SNAPSHOT, r27002+6-a9f58dddb7
 -----------------------------------------------------
root@wndap620:~# 
```
Device is reachable over LAN and can login through ssh. 

Tested-by: Christian Lamparter <chunkeey@gmail.com> (wndap620+wnap660)

namiltd commented:
I have doubts about the detection of chip rtl8363sb. 
Axzez Interceptor Carrier Board has rtl8367rb chip: https://pipci.jeffgeerling.com/images/board-axzez-interceptor.jpeg
According to https://www.axzez.com/forum/interceptor-carrier-board/rtl8367rb-chip-detection-failed-err-19-interceptor-os-22-08-12-29-img chip rtl8367rb has the same chip_num and chip_ver

namiltd commented:
According to https://axzez-customer-deliverables.nyc3.cdn.digitaloceanspaces.com/linux-patches-170.tar.gz sthe RTL8367RB switch is detected identically to the RTL8363SB (chip_ver:1000, chip_num:6000). So I'm removing the commit https://github.com/openwrt/openwrt/commit/ec288d7564e0e513630ba0bfc396eabad9eb184c.

namiltd commented:
The code is ready for review.

namiltd commented:
Temporary cherry-picked https://github.com/openwrt/openwrt/pull/16051/commits/c2d1b05ae7a52f9db429455ac4267a4c417ffa04

namiltd commented:
The code is ready for review.

@hauke What do you think, is it better to use a simpler code without rlvid or the code with RTL8367B_CHIP_UNKNOWN_RLVID0 and RTL8367B_CHIP_UNKNOWN_RLVID1 but they are not currently used in any of the routers. Do "spare" types make sense?

namiltd commented:
@hauke Everything is ready, the code does not include rlvid analysis which was dead code, which makes the resulting code simpler.

namiltd commented:
@hauke Everything is ready, the code does not include rlvid analysis which was dead code, which makes the resulting code simpler.

namiltd committed:
kernel: rtl8367b: store chip in smi->rtl8367b_chip

Store the chip type in smi->rtl8367b_chip

Signed-off-by: Mieczyslaw Nalewaj <namiltd@yahoo.com>
Link: https://github.com/openwrt/openwrt/pull/15749
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

namiltd commented:
@chunkeey Can you test again but with an additional patch?: [debug0.patch](https://github.com/user-attachments/files/16488885/debug0.patch)
